# T81-558: Applications of Deep Neural Networks
**Module 2: Python for Machine Learning**
* Instructor: [Jeff Heaton](https://sites.wustl.edu/jeffheaton/), McKelvey School of Engineering, [Washington University in St. Louis](https://engineering.wustl.edu/Programs/Pages/default.aspx)
* For more information visit the [class website](https://sites.wustl.edu/jeffheaton/t81-558/).

# Module Video Material

Main video lecture:

* [Part 2.1: Introduction to Pandas](https://www.youtube.com/watch?v=Bj2m6hvRoNk&index=6&list=PLjy4p-07OYzulelvJ5KVaT2pDlxivl_BN)
* [Part 2.2: Categorical Values](https://www.youtube.com/watch?v=WCXzchgxi9c&list=PLjy4p-07OYzulelvJ5KVaT2pDlxivl_BN)
* [Part 2.3: Grouping, Sorting, and Shuffling](https://www.youtube.com/watch?v=WCXzchgxi9c&list=PLjy4p-07OYzulelvJ5KVaT2pDlxivl_BN)
* [Part 2.4: Apply and Map](https://www.youtube.com/watch?v=eZGunTjrHyA&list=PLjy4p-07OYzulelvJ5KVaT2pDlxivl_BN)
* [Part 2.5: Feature Engineering](https://www.youtube.com/watch?v=eZGunTjrHyA&list=PLjy4p-07OYzulelvJ5KVaT2pDlxivl_BN)

# Part 2.1: Introduction to Pandas

[Pandas](http://pandas.pydata.org/) is an open source library providing high-performance, easy-to-use data structures and data analysis tools for the Python programming language.  It is based on the [dataframe](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html) concept found in the [R programming language](https://www.r-project.org/about.html).  For this class, Pandas will be the primary means by which data is manipulated in conjunction with neural networks.

The dataframe is a key component of Pandas.  We will use it to access the [auto-mpg dataset](https://archive.ics.uci.edu/ml/datasets/Auto+MPG).  This dataset can be found on the UCI machine learning repository.  For this class we will use a version of the Auto MPG dataset where I added column headers.  You can find my version [here](https://raw.githubusercontent.com/jeffheaton/t81_558_deep_learning/master/data/auto-mpg.csv).

This dataset was taken from the StatLib library which is maintained at Carnegie Mellon University. The dataset was used in the 1983 American Statistical Association Exposition.  It contains data for 398 cars, including [mpg](https://en.wikipedia.org/wiki/Fuel_economy_in_automobiles), [cylinders](https://en.wikipedia.org/wiki/Cylinder_(engine)), [displacement](https://en.wikipedia.org/wiki/Engine_displacement), [horsepower](https://en.wikipedia.org/wiki/Horsepower) , weight, acceleration, model year, origin and the car's name.

The following code loads the MPG dataset into a dataframe:

In [1]:
# Simple dataframe
import os
import pandas as pd

df = pd.read_csv("https://data.heatonresearch.com/data/t81-558/auto-mpg.csv")
print(df[0:5])

    mpg  cylinders  displacement horsepower  weight  acceleration  year  \
0  18.0          8         307.0        130    3504          12.0    70   
1  15.0          8         350.0        165    3693          11.5    70   
2  18.0          8         318.0        150    3436          11.0    70   
3  16.0          8         304.0        150    3433          12.0    70   
4  17.0          8         302.0        140    3449          10.5    70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino  


The **display** function provides a cleaner display than simply printing the dataframe.

In [2]:
display(df[0:5])

mpg  cylinders  displacement horsepower  weight  acceleration  year  \
0  18.0          8         307.0        130    3504          12.0    70   
1  15.0          8         350.0        165    3693          11.5    70   
2  18.0          8         318.0        150    3436          11.0    70   
3  16.0          8         304.0        150    3433          12.0    70   
4  17.0          8         302.0        140    3449          10.5    70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

It is possible to generate a second dataframe to display statistical information about the first dataframe.

In [3]:
# Strip non-numerics
df = df.select_dtypes(include=['int', 'float'])

headers = list(df.columns.values)
fields = []

for field in headers:
    fields.append({
        'name' : field,
        'mean': df[field].mean(),
        'var': df[field].var(),
        'sdev': df[field].std()
    })

for field in fields:
    print(field)

{'name': 'mpg', 'mean': 23.514572864321607, 'var': 61.089610774274405, 'sdev': 7.815984312565782}
{'name': 'cylinders', 'mean': 5.454773869346734, 'var': 2.893415439920003, 'sdev': 1.7010042445332119}
{'name': 'displacement', 'mean': 193.42587939698493, 'var': 10872.199152247384, 'sdev': 104.26983817119591}
{'name': 'weight', 'mean': 2970.424623115578, 'var': 717140.9905256763, 'sdev': 846.8417741973268}
{'name': 'acceleration', 'mean': 15.568090452261307, 'var': 7.604848233611383, 'sdev': 2.757688929812676}
{'name': 'year', 'mean': 76.01005025125629, 'var': 13.672442818627143, 'sdev': 3.697626646732623}
{'name': 'origin', 'mean': 1.5728643216080402, 'var': 0.6432920268850549, 'sdev': 0.8020548777266148}


Converting this to a dataframe allows for a nicer display.

In [4]:
df2 = pd.DataFrame(fields)
display(df2)

mean          name        sdev            var
0    23.514573           mpg    7.815984      61.089611
1     5.454774     cylinders    1.701004       2.893415
2   193.425879  displacement  104.269838   10872.199152
3  2970.424623        weight  846.841774  717140.990526
4    15.568090  acceleration    2.757689       7.604848
5    76.010050          year    3.697627      13.672443
6     1.572864        origin    0.802055       0.643292

## Missing Values

Missing values are a reality of machine learning.  Ideally every row of data will have values for all columns.  However, this is rarely the case.  Most of the values are present in the MPG database.  However, there are missing values in the horsepower column.  A common practice is to replace missing values with the median value for that column.  The median is calculated as described [here](https://www.mathsisfun.com/median.html).  The following code replaces any NA values in horsepower with the median:

In [5]:
import os
import pandas as pd

df = pd.read_csv(
    "https://data.heatonresearch.com/data/t81-558/auto-mpg.csv", 
    na_values=['NA', '?'])
print(f"horsepower has na? {pd.isnull(df['horsepower']).values.any()}")
    
print("Filling missing values...")
med = df['horsepower'].median()
df['horsepower'] = df['horsepower'].fillna(med)
# df = df.dropna() # you can also simply drop NA values
                 
print(f"horsepower has na? {pd.isnull(df['horsepower']).values.any()}")

horsepower has na? True
Filling missing values...
horsepower has na? False


# Dealing with Outliers

Outliers are values that are unusually high or low.  Sometimes outliers are simply errors, this is a result of [observation error](https://en.wikipedia.org/wiki/Observational_error).  Outliers can also be truly large or small values that may be difficult to account for. Outliers are usually defined as a value that is several standard deviations from the mean.  The following function can be used to remove such values.    

In [6]:
# Remove all rows where the specified column is +/- sd standard deviations
def remove_outliers(df, name, sd):
    drop_rows = df.index[(np.abs(df[name] - df[name].mean())
                          >= (sd * df[name].std()))]
    df.drop(drop_rows, axis=0, inplace=True)

The code below will drop every row from the Auto MPG dataset where the horsepower is more than 2 standard deviations above or below the mean.

In [7]:
import pandas as pd
import os
import numpy as np
from sklearn import metrics
from scipy.stats import zscore

df = pd.read_csv(
    "https://data.heatonresearch.com/data/t81-558/auto-mpg.csv",
    na_values=['NA','?'])

# create feature vector
med = df['horsepower'].median()
df['horsepower'] = df['horsepower'].fillna(med)

# Drop the name column
df.drop('name',1,inplace=True)

# Drop outliers in horsepower
print("Length before MPG outliers dropped: {}".format(len(df)))
remove_outliers(df,'mpg',2)
print("Length after MPG outliers dropped: {}".format(len(df)))

display(df[0:5])

Length before MPG outliers dropped: 398
Length after MPG outliers dropped: 388


mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0  18.0          8         307.0       130.0    3504          12.0    70   
1  15.0          8         350.0       165.0    3693          11.5    70   
2  18.0          8         318.0       150.0    3436          11.0    70   
3  16.0          8         304.0       150.0    3433          12.0    70   
4  17.0          8         302.0       140.0    3449          10.5    70   

   origin  
0       1  
1       1  
2       1  
3       1  
4       1

## Dropping Fields

Some fields are of no value to the neural network and can be dropped.  The following code removes the name column from the MPG dataset.

In [8]:
import os
import pandas as pd

df = pd.read_csv(
    "https://data.heatonresearch.com/data/t81-558/auto-mpg.csv",
    na_values=['NA','?'])

print(f"Before drop: {list(df.columns)}")
df.drop('name', 1, inplace=True)
print(f"After drop: {list(df.columns)}")

Before drop: ['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', 'year', 'origin', 'name']
After drop: ['mpg', 'cylinders', 'displacement', 'horsepower', 'weight', 'acceleration', 'year', 'origin']


## Concatenating Rows and Columns
Rows and columns can be concatenated together to form new data frames.  The code below creates a new dataframe from name and horsepower from the Auto MPG dataset.  This is done by concatenating two columns together.

In [9]:
# Create a new dataframe from name and horsepower

import os
import pandas as pd

df = pd.read_csv(
    "https://data.heatonresearch.com/data/t81-558/auto-mpg.csv",
    na_values=['NA','?'])

col_horsepower = df['horsepower']
col_name = df['name']
result = pd.concat([col_name, col_horsepower], axis=1)
display(result[0:5])

name  horsepower
0  chevrolet chevelle malibu       130.0
1          buick skylark 320       165.0
2         plymouth satellite       150.0
3              amc rebel sst       150.0
4                ford torino       140.0

The **concat** function can also concatenate two rows together.  This code concatenates the first 2 rows and the last 2 rows of the Auto MPG dataset.

In [10]:
# Create a new dataframe from first 2 rows and last 2 rows

import os
import pandas as pd

df = pd.read_csv(
    "https://data.heatonresearch.com/data/t81-558/auto-mpg.csv",
    na_values=['NA','?'])

result = pd.concat([df[0:2],df[-2:]], axis=0)
display(result)

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0    18.0          8         307.0       130.0    3504          12.0    70   
1    15.0          8         350.0       165.0    3693          11.5    70   
396  28.0          4         120.0        79.0    2625          18.6    82   
397  31.0          4         119.0        82.0    2720          19.4    82   

     origin                       name  
0         1  chevrolet chevelle malibu  
1         1          buick skylark 320  
396       1                ford ranger  
397       1                 chevy s-10

## Training and Validation

It is very important that we evaluate a machine learning model based on its ability to predict data that it has never seen before.  Because of this we often divide the training data into a validation and training set.  The machine learning model will learn from the training data, but ultimately be evaluated based on the validation data.

* **Training Data** - **In Sample Data** - The data that the machine learning model was fit to/created from. 
* **Validation Data** - **Out of Sample Data** - The data that the machine learning model is evaluated upon after it is fit to the training data.

There are two predominant means of dealing with training and validation data:

* **Training/Validation Split** - The data are split according to some ratio between a training and validation (hold-out) set.  Common ratios are 80% training and 20% validation.
* **K-Fold Cross Validation** - The data are split into a number of folds and models.  Because a number of models equal to the folds is created out-of-sample predictions can be generated for the entire dataset.

The code below performs a split of the MPG data into a training and validation set.  The training set uses 80% of the data and the validation set uses 20%.

The following image shows how a model is trained on 80% of the data and then validated against the remaining 20%.

![Training and Validation](https://raw.githubusercontent.com/jeffheaton/t81_558_deep_learning/master/images/class_1_train_val.png "Training and Validation")


In [11]:
import os
import pandas as pd
import numpy as np

df = pd.read_csv(
    "https://data.heatonresearch.com/data/t81-558/auto-mpg.csv",
    na_values=['NA','?'])

df = df.reindex(np.random.permutation(df.index)) # Usually a good idea to shuffle
mask = np.random.rand(len(df)) < 0.8
trainDF = pd.DataFrame(df[mask])
validationDF = pd.DataFrame(df[~mask])

print(f"Training DF: {len(trainDF)}")
print(f"Validation DF: {len(validationDF)}")

Training DF: 313
Validation DF: 85


### Converting a Dataframe to a Matrix

Neural networks do not directly operate on Python dataframes.  A neural network requires a numeric matrix.  The **values** property of a dataframe is used to convert to a matrix.

In [12]:
df.values

array([[13.0, 8, 318.0, ..., 76, 1, 'plymouth volare premier v8'],
       [24.0, 4, 119.0, ..., 75, 3, 'datsun 710'],
       [31.3, 4, 120.0, ..., 80, 3, 'mazda 626'],
       ...,
       [15.0, 8, 390.0, ..., 70, 1, 'amc ambassador dpl'],
       [14.0, 8, 350.0, ..., 71, 1, 'chevrolet impala'],
       [18.0, 6, 225.0, ..., 74, 1, 'plymouth satellite sebring']],
      dtype=object)

You might wish to only convert some of the columns, to leave out the name column, use the following code.

In [13]:
df[['mpg', 'cylinders', 'displacement', 'horsepower', 'weight',
       'acceleration', 'year', 'origin']].values

array([[ 13. ,   8. , 318. , ...,  13.2,  76. ,   1. ],
       [ 24. ,   4. , 119. , ...,  17. ,  75. ,   3. ],
       [ 31.3,   4. , 120. , ...,  17.5,  80. ,   3. ],
       ...,
       [ 15. ,   8. , 390. , ...,   8.5,  70. ,   1. ],
       [ 14. ,   8. , 350. , ...,  12. ,  71. ,   1. ],
       [ 18. ,   6. , 225. , ...,  16.5,  74. ,   1. ]])

## Saving a Dataframe to CSV

Many of the assignments in this course will require that you save a dataframe to submit to the instructor.  The following code performs a shuffle and then saves a new copy.

In [14]:
import os
import pandas as pd
import numpy as np

path = "./data/"

df = pd.read_csv(
    "https://data.heatonresearch.com/data/t81-558/auto-mpg.csv",
    na_values=['NA','?'])

filename_write = os.path.join(path, "auto-mpg-shuffle.csv")
df = df.reindex(np.random.permutation(df.index))
df.to_csv(filename_write, index=False) # Specify index = false to not write row numbers
print("Done")

Done


## Saving a Dataframe to Pickle

CSV files are text and can be used by a variety of software programs.  However, they do take longer to generate and can sometimes lose small amounts of precision in the conversion.  Another format is [Pickle](https://docs.python.org/3/library/pickle.html).  Generally you will output to CSV because it is very compatible, even outside of Python.  The code below stores the Dataframe to Pickle.

In [15]:
import os
import pandas as pd
import numpy as np
import pickle

path = "./data/"

df = pd.read_csv(
    "https://data.heatonresearch.com/data/t81-558/auto-mpg.csv",
    na_values=['NA','?'])

filename_write = os.path.join(path, "auto-mpg-shuffle.pkl")
df = df.reindex(np.random.permutation(df.index))

with open(filename_write,"wb") as fp:
    pickle.dump(df, fp)

print("Done")

Done


Loading the pickle file back into memory is accomplished by the following lines of code.  Notice that the index numbers are still jumbled from the previous shuffle?  Loading the CSV rebuilt (in the previous step) did not preserve these values.

In [16]:
import os
import pandas as pd
import numpy as np
import pickle

path = "./data/"

df = pd.read_csv(
    "https://data.heatonresearch.com/data/t81-558/auto-mpg.csv",
    na_values=['NA','?'])

filename_read = os.path.join(path, "auto-mpg-shuffle.pkl")

with open(filename_write,"rb") as fp:
    df = pickle.load(fp)

display(df[0:5])

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
355  33.7          4         107.0        75.0    2210          14.4    81   
15   22.0          6         198.0        95.0    2833          15.5    70   
57   24.0          4         113.0        95.0    2278          15.5    72   
246  32.8          4          78.0        52.0    1985          19.4    78   
194  22.5          6         232.0        90.0    3085          17.6    76   

     origin                   name  
355       3          honda prelude  
15        1        plymouth duster  
57        3  toyota corona hardtop  
246       3       mazda glc deluxe  
194       1             amc hornet

# Part 2.2: Categorical and Continuous Values

Neural networks require their input to be a fixed number of columns.  This is very similar to spreadsheet data.  This input must be completely numeric.  

It is important to represent the data in a way that the neural network can train from it.  In class 6, we will see even more ways to preprocess data.  For now, we will look at several of the most basic ways to transform data for a neural network.

Before we look at specific ways to preprocess data, it is important to consider four basic types of data, as defined by [Stanley Smith Stevens](https://en.wikipedia.org/wiki/Stanley_Smith_Stevens).  These are commonly referred to as the [levels of measure](https://en.wikipedia.org/wiki/Level_of_measurement):

* Character Data (strings)
    * **Nominal** - Individual discrete items, no order. For example: color, zip code, shape.
    * **Ordinal** - Individual discrete items that can be ordered.  For example: grade level, job title, Starbucks(tm) coffee size (tall, vente, grande) 
* Numeric Data
    * **Interval** - Numeric values, no defined start.  For example, temperature.  You would never say "yesterday was twice as hot as today".
    * **Ratio** - Numeric values, clearly defined start.  For example, speed.  You would say that "The first car is going twice as fast as the second."

### Encoding Continuous Values

One common transformation is to normalize the inputs.  It is sometimes valuable to normalization numeric inputs to be put in a standard form so that two values can easily be compared.  Consider if a friend told you that he received a $10 discount.  Is this a good deal?  Maybe.  But the value is not normalized.  If your friend purchased a car, then the discount is not that good.  If your friend purchased dinner, this is a very good discount!

Percentages are a very common form of normalization.  If your friend tells you they got 10% off, we know that this is a better discount than 5%.  It does not matter how much the purchase price was.  One very common machine learning normalization is the Z-Score:

$z = \frac{x - \mu}{\sigma} $

To calculate the Z-Score you need to also calculate the mean($\mu$) and the standard deviation ($\sigma$).  The mean is calculated as follows:

$\mu = \bar{x} = \frac{x_1+x_2+\cdots +x_n}{n}$

The standard deviation is calculated as follows:

$\sigma = \sqrt{\frac{1}{N} \sum_{i=1}^N (x_i - \mu)^2}, {\rm \ \ where\ \ } \mu = \frac{1}{N} \sum_{i=1}^N x_i$

The following Python code replaces the mpg with a z-score.  Cars with average MPG will be near zero, above zero is above average, and below zero is below average.  Z-Scores above/below -3/3 are very rare, these are outliers.

In [17]:
import os
import pandas as pd
from scipy.stats import zscore

df = pd.read_csv(
    "https://data.heatonresearch.com/data/t81-558/auto-mpg.csv",
    na_values=['NA','?'])

df['mpg'] = zscore(df['mpg'])
display(df[0:5])

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0 -0.706439          8         307.0       130.0    3504          12.0    70   
1 -1.090751          8         350.0       165.0    3693          11.5    70   
2 -0.706439          8         318.0       150.0    3436          11.0    70   
3 -0.962647          8         304.0       150.0    3433          12.0    70   
4 -0.834543          8         302.0       140.0    3449          10.5    70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

### Encoding Categorical Values as Dummies
The classic means of encoding categorical values is to make them dummy variables.  This is also called one-hot-encoding.  Consider the following data set.

In [18]:
import pandas as pd

df = pd.read_csv(
    "https://data.heatonresearch.com/data/t81-558/jh-simple-dataset.csv",
    na_values=['NA','?'])

display(df[1:5])

id job area   income     aspect  subscriptions  dist_healthy  save_rate  \
1   2  kd    c  60369.0  18.625000              2      7.766643         59   
2   3  pe    c  55126.0  34.766667              1      3.632069          6   
3   4  11    c  51690.0  15.808333              1      5.372942         16   
4   5  kl    d  28347.0  40.941667              3      3.822477         20   

   dist_unhealthy  age  pop_dense  retail_dense     crime product  
1        6.805396   51   0.874016      0.342520  0.400809       c  
2       13.671772   44   0.944882      0.724409  0.207723       b  
3        4.333286   50   0.889764      0.444882  0.361216       b  
4        5.967121   38   0.744094      0.661417  0.068033       a

In [19]:
areas = list(df['area'].unique())
print(f'Number of areas: {len(areas)}')
print(f'Areas: {areas}')

Number of areas: 4
Areas: ['c', 'd', 'a', 'b']


There are four unique values in the areas column.  To encode these to dummy variables we would use four columns, each of which would represent one of the areas.  For each row, one column would have a value of one, the rest zeros.  This is why this type of encoding is sometimes called one-hot encoding.  The following code shows how you might encode the values "a" through "d".  The value A becomes [1,0,0,0] and the value B becomes [0,1,0,0].

In [20]:
dummies = pd.get_dummies(['a','b','c','d'])
print(dummies)

   a  b  c  d
0  1  0  0  0
1  0  1  0  0
2  0  0  1  0
3  0  0  0  1


To encode the "area" column, we use the following.

In [21]:
dummies = pd.get_dummies(df['area'])
print(dummies[0:10]) # Just show the first 10

   a  b  c  d
0  0  0  1  0
1  0  0  1  0
2  0  0  1  0
3  0  0  1  0
4  0  0  0  1
5  0  0  1  0
6  0  0  0  1
7  1  0  0  0
8  0  0  1  0
9  1  0  0  0


Since we will merge the above values back into the dataframe, it is helpful to know that these dummies belongs to the original area column.  We can rename the dummy values like this.

In [22]:
name = "area"    

for x in dummies.columns:
    dummy_name = f"{name}-{x}"
    df[dummy_name] = dummies[x]

Displaying select columns from the dataset we can see the dummy variables added.

In [23]:
display(df[0:10][['id','job','area','income','area-a',
                  'area-b','area-c','area-d']])

id job area   income  area-a  area-b  area-c  area-d
0   1  vv    c  50876.0       0       0       1       0
1   2  kd    c  60369.0       0       0       1       0
2   3  pe    c  55126.0       0       0       1       0
3   4  11    c  51690.0       0       0       1       0
4   5  kl    d  28347.0       0       0       0       1
5   6  e2    c  70854.0       0       0       1       0
6   7  kl    d  38726.0       0       0       0       1
7   8  nb    a  55162.0       1       0       0       0
8   9  al    c  67311.0       0       0       1       0
9  10  pe    a  63344.0       1       0       0       0

Usually, you will remove the original column ('area'), because it is the goal to get the dataframe to be entirely numeric for the neural network.

In [24]:
df.drop('area', axis=1, inplace=True)
display(df[0:10][['id','job','income','area-a',
                  'area-b','area-c','area-d']])

id job   income  area-a  area-b  area-c  area-d
0   1  vv  50876.0       0       0       1       0
1   2  kd  60369.0       0       0       1       0
2   3  pe  55126.0       0       0       1       0
3   4  11  51690.0       0       0       1       0
4   5  kl  28347.0       0       0       0       1
5   6  e2  70854.0       0       0       1       0
6   7  kl  38726.0       0       0       0       1
7   8  nb  55162.0       1       0       0       0
8   9  al  67311.0       0       0       1       0
9  10  pe  63344.0       1       0       0       0

This entire process can be contained in the following function:

In [25]:
def encode_text_dummy(df, name):
    dummies = pd.get_dummies(df[name])
    for x in dummies.columns:
        dummy_name = f"{name}-{x}"
        df[dummy_name] = dummies[x]
    df.drop(name, axis=1, inplace=True)

### Target Encoding for Categoricals

Target encoding can sometimes increase the predictive power of a machine learning model.  However, it also greatly increases the risk of overfitting.  Because of this risk, care must be take if you are going to use this method.  It is a popular technique for Kaggle competitions.  

Generally, target encoding can only be used on a categorical feature when the output of the machine learning model is numeric (regression).

The concept of target encoding is actually very simple.  For each value 

In [26]:
# Create a small sample dataset
import pandas as pd
import numpy as np

np.random.seed(43)
df = pd.DataFrame({
    'cont_9': np.random.rand(10)*100,
    'cat_0': ['dog'] * 5 + ['cat'] * 5,
    'cat_1': ['wolf'] * 9 + ['tiger'] * 1,
    'y': [1, 0, 1, 1, 1, 1, 0, 0, 0, 0]
})

display(df)

cont_9 cat_0  cat_1  y
0  11.505457   dog   wolf  1
1  60.906654   dog   wolf  0
2  13.339096   dog   wolf  1
3  24.058962   dog   wolf  1
4  32.713906   dog   wolf  1
5  85.913749   cat   wolf  1
6  66.609021   cat   wolf  0
7  54.116221   cat   wolf  0
8   2.901382   cat   wolf  0
9  73.374830   cat  tiger  0

Rather than creating dummy variables for dog and cat, we would like to change it to a number. We could just use 0 for cat, 1 for dog.  However, we can encode more information than just that.  The simple 0 or 1 would also only work for one animal.  Consider what the mean target value is for cat and dog.

In [27]:
means0 = df.groupby('cat_0')['y'].mean().to_dict()
means0

{'cat': 0.2, 'dog': 0.8}

The danger is that we are now using the target value for training.  This will potentially overfit.  The possibility of overfitting is even greater if there are a small number of a particular category.  To prevent this from happening, we use a weighting factor.  The stronger the weight the more than categories with a small number of values will tend towards the overall average of y, which is calculated as follows.

In [28]:
df['y'].mean()

0.5

The complete function for target encoding is given here.

In [29]:
# Source: https://maxhalford.github.io/blog/target-encoding-done-the-right-way/
def calc_smooth_mean(df1, df2, cat_name, target, weight):
    # Compute the global mean
    mean = df[target].mean()

    # Compute the number of values and the mean of each group
    agg = df.groupby(cat_name)[target].agg(['count', 'mean'])
    counts = agg['count']
    means = agg['mean']

    # Compute the "smoothed" means
    smooth = (counts * means + weight * mean) / (counts + weight)

    # Replace each value by the according smoothed mean
    if df2 is None:
        return df1[cat_name].map(smooth)
    else:
        return df1[cat_name].map(smooth),df2[cat_name].map(smooth.to_dict())

The following code encodes these two categories.

In [30]:
WEIGHT = 5
df['cat_0_enc'] = calc_smooth_mean(df1=df, df2=None, cat_name='cat_0', target='y', weight=WEIGHT)
df['cat_1_enc'] = calc_smooth_mean(df1=df, df2=None, cat_name='cat_1', target='y', weight=WEIGHT)

In [31]:
display(df)

cont_9 cat_0  cat_1  y  cat_0_enc  cat_1_enc
0  11.505457   dog   wolf  1       0.65   0.535714
1  60.906654   dog   wolf  0       0.65   0.535714
2  13.339096   dog   wolf  1       0.65   0.535714
3  24.058962   dog   wolf  1       0.65   0.535714
4  32.713906   dog   wolf  1       0.65   0.535714
5  85.913749   cat   wolf  1       0.35   0.535714
6  66.609021   cat   wolf  0       0.35   0.535714
7  54.116221   cat   wolf  0       0.35   0.535714
8   2.901382   cat   wolf  0       0.35   0.535714
9  73.374830   cat  tiger  0       0.35   0.416667

### Encoding Categorical Values as Ordinal

Typically categoricals will be encoded as dummy variables.  However, there might be other techniques to convert categoricals to numeric. Any time there is an order to the categoricals, a number should be used.  Consider if you had a categorical that described the current education level of an individual.   

* Kindergarten (0)
* First Grade (1)
* Second Grade (2)
* Third Grade (3)
* Fourth Grade (4)
* Fifth Grade (5)
* Sixth Grade (6)
* Seventh Grade (7)
* Eighth Grade (8)
* High School Freshman (9)
* High School Sophomore (10)
* High School Junior (11)
* High School Senior (12)
* College Freshman (13)
* College Sophomore (14)
* College Junior (15)
* College Senior (16)
* Graduate Student (17)
* PhD Candidate (18)
* Doctorate (19)
* Post Doctorate (20)

The above list has 21 levels.  This would take 21 dummy variables. However, simply encoding this to dummies would lose the order information.  Perhaps the easiest approach would be to assign simply number them and assign the category a single number that is equal to the value in parenthesis above.  However, we might be able to do even better.  Graduate student is likely more than a year, so you might increase more than just one value.  

# Part 2.3: Grouping, Sorting, and Shuffling  

### Shuffling a Dataset
The following code is used to shuffle and reindex a data set.  A random seed can be used to produce a consistent shuffling of the data set.

In [32]:
import os
import pandas as pd

df = pd.read_csv(
    "https://data.heatonresearch.com/data/t81-558/auto-mpg.csv", 
    na_values=['NA', '?'])

#np.random.seed(42) # Uncomment this line to get the same shuffle each time
df = df.reindex(np.random.permutation(df.index))
df.reset_index(inplace=True, drop=True)
display(df[0:10])

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0  25.4          6         168.0       116.0    2900          12.6    81   
1  36.4          5         121.0        67.0    2950          19.9    80   
2  23.9          8         260.0        90.0    3420          22.2    79   
3  30.0          4          98.0        68.0    2155          16.5    78   
4  31.0          4         119.0        82.0    2720          19.4    82   
5  34.2          4         105.0        70.0    2200          13.2    79   
6  20.2          6         200.0        85.0    2965          15.8    78   
7  26.0          4         122.0        80.0    2451          16.5    74   
8  15.0          8         350.0       145.0    4440          14.0    75   
9  39.1          4          79.0        58.0    1755          16.9    81   

   origin                               name  
0       3                    toyota cressida  
1       2                audi 5000s (diesel)  
2       1  oldsmobile cutlass salon brougham  
3       1                 chevrolet chevette  
4       1                         chevy s-10  
5       1                   plymouth horizon  
6       1               ford fairmont (auto)  
7       1                         ford pinto  
8       1                  chevrolet bel air  
9       3                     toyota starlet

### Sorting a Data Set

Data sets can also be sorted.  This code sorts the MPG dataset by name and displays the first car.

In [33]:
import os
import pandas as pd

df = pd.read_csv(
    "https://data.heatonresearch.com/data/t81-558/auto-mpg.csv", 
    na_values=['NA', '?'])

df = df.sort_values(by='name', ascending=True)
print(f"The first car is: {df['name'].iloc[0]}")
display(df[0:5])

The first car is: amc ambassador brougham


mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
96   13.0          8         360.0       175.0    3821          11.0    73   
9    15.0          8         390.0       190.0    3850           8.5    70   
66   17.0          8         304.0       150.0    3672          11.5    72   
315  24.3          4         151.0        90.0    3003          20.1    80   
257  19.4          6         232.0        90.0    3210          17.2    78   

     origin                     name  
96        1  amc ambassador brougham  
9         1       amc ambassador dpl  
66        1       amc ambassador sst  
315       1              amc concord  
257       1              amc concord

### Grouping a Data Set

Grouping is a common operation on data sets.  In SQL, this operation is referred to as "GROUP BY".  Grouping is used to summarize data.  Because of this summarization the row could will either stay the same or more likely shrink after a grouping is applied.

The Auto MPG dataset is used to demonstrate grouping.

In [34]:
import os
import pandas as pd

df = pd.read_csv(
    "https://data.heatonresearch.com/data/t81-558/auto-mpg.csv", 
    na_values=['NA', '?'])
display(df[0:5])

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0  18.0          8         307.0       130.0    3504          12.0    70   
1  15.0          8         350.0       165.0    3693          11.5    70   
2  18.0          8         318.0       150.0    3436          11.0    70   
3  16.0          8         304.0       150.0    3433          12.0    70   
4  17.0          8         302.0       140.0    3449          10.5    70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

The above data set can be used with group to perform summaries.  For example, the following code will group cylinders by the average (mean).  This code will provide the grouping.  In addition to mean, other aggregating functions, such as **sum** or **count** can be used. 

In [35]:
g = df.groupby('cylinders')['mpg'].mean()
g

cylinders
3    20.550000
4    29.286765
5    27.366667
6    19.985714
8    14.963107
Name: mpg, dtype: float64

It might be useful to have these **mean** values as a dictionary.

In [36]:
d = g.to_dict()
d

{3: 20.55,
 4: 29.28676470588236,
 5: 27.366666666666664,
 6: 19.985714285714284,
 8: 14.963106796116508}

This allows you to quickly access an individual element, such as to lookup the mean for 6 cylinders.  This is used in target encoding, which is presented in this module.

In [37]:
d[6]

19.985714285714284

The code below shows how to count the number of rows that match each cylinder count.

In [38]:
df.groupby('cylinders')['mpg'].count().to_dict()

{3: 4, 4: 204, 5: 3, 6: 84, 8: 103}

# Part 2.4: Apply and Map

The **apply** and **map** functions can also be applied to Pandas **dataframes**.

### Using Map with Dataframes

In [39]:
import os
import pandas as pd
import numpy as np

df = pd.read_csv(
    "https://data.heatonresearch.com/data/t81-558/auto-mpg.csv", 
    na_values=['NA', '?'])

display(df[0:10])

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0  18.0          8         307.0       130.0    3504          12.0    70   
1  15.0          8         350.0       165.0    3693          11.5    70   
2  18.0          8         318.0       150.0    3436          11.0    70   
3  16.0          8         304.0       150.0    3433          12.0    70   
4  17.0          8         302.0       140.0    3449          10.5    70   
5  15.0          8         429.0       198.0    4341          10.0    70   
6  14.0          8         454.0       220.0    4354           9.0    70   
7  14.0          8         440.0       215.0    4312           8.5    70   
8  14.0          8         455.0       225.0    4425          10.0    70   
9  15.0          8         390.0       190.0    3850           8.5    70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino  
5       1           ford galaxie 500  
6       1           chevrolet impala  
7       1          plymouth fury iii  
8       1           pontiac catalina  
9       1         amc ambassador dpl

In [40]:
df['origin_name'] = df['origin'].map({1: 'North America', 2: 'Europe', 3: 'Asia'})
display(df[0:50])

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0   18.0          8         307.0       130.0    3504          12.0    70   
1   15.0          8         350.0       165.0    3693          11.5    70   
2   18.0          8         318.0       150.0    3436          11.0    70   
3   16.0          8         304.0       150.0    3433          12.0    70   
4   17.0          8         302.0       140.0    3449          10.5    70   
5   15.0          8         429.0       198.0    4341          10.0    70   
6   14.0          8         454.0       220.0    4354           9.0    70   
7   14.0          8         440.0       215.0    4312           8.5    70   
8   14.0          8         455.0       225.0    4425          10.0    70   
9   15.0          8         390.0       190.0    3850           8.5    70   
10  15.0          8         383.0       170.0    3563          10.0    70   
11  14.0          8         340.0       160.0    3609           8.0    70   
12  15.0          8         400.0       150.0    3761           9.5    70   
13  14.0          8         455.0       225.0    3086          10.0    70   
14  24.0          4         113.0        95.0    2372          15.0    70   
15  22.0          6         198.0        95.0    2833          15.5    70   
16  18.0          6         199.0        97.0    2774          15.5    70   
17  21.0          6         200.0        85.0    2587          16.0    70   
18  27.0          4          97.0        88.0    2130          14.5    70   
19  26.0          4          97.0        46.0    1835          20.5    70   
20  25.0          4         110.0        87.0    2672          17.5    70   
21  24.0          4         107.0        90.0    2430          14.5    70   
22  25.0          4         104.0        95.0    2375          17.5    70   
23  26.0          4         121.0       113.0    2234          12.5    70   
24  21.0          6         199.0        90.0    2648          15.0    70   
25  10.0          8         360.0       215.0    4615          14.0    70   
26  10.0          8         307.0       200.0    4376          15.0    70   
27  11.0          8         318.0       210.0    4382          13.5    70   
28   9.0          8         304.0       193.0    4732          18.5    70   
29  27.0          4          97.0        88.0    2130          14.5    71   
30  28.0          4         140.0        90.0    2264          15.5    71   
31  25.0          4         113.0        95.0    2228          14.0    71   
32  25.0          4          98.0         NaN    2046          19.0    71   
33  19.0          6         232.0       100.0    2634          13.0    71   
34  16.0          6         225.0       105.0    3439          15.5    71   
35  17.0          6         250.0       100.0    3329          15.5    71   
36  19.0          6         250.0        88.0    3302          15.5    71   
37  18.0          6         232.0       100.0    3288          15.5    71   
38  14.0          8         350.0       165.0    4209          12.0    71   
39  14.0          8         400.0       175.0    4464          11.5    71   
40  14.0          8         351.0       153.0    4154          13.5    71   
41  14.0          8         318.0       150.0    4096          13.0    71   
42  12.0          8         383.0       180.0    4955          11.5    71   
43  13.0          8         400.0       170.0    4746          12.0    71   
44  13.0          8         400.0       175.0    5140          12.0    71   
45  18.0          6         258.0       110.0    2962          13.5    71   
46  22.0          4         140.0        72.0    2408          19.0    71   
47  19.0          6         250.0       100.0    3282          15.0    71   
48  18.0          6         250.0        88.0    3139          14.5    71   
49  23.0          4         122.0        86.0    2220          14.0    71   

    origin                          name    origin_name  
0        1     chev

### Using Apply with Dataframes

If the **apply** function is directly executed on the data frame, the lambda function is called once per column or row, depending on the value of axis.  For axis = 1, rows are used. 

The following code calculates a series called **efficiency** that is the **displacement** divided by **horsepower**. 

In [41]:
effi = df.apply(lambda x: x['displacement']/x['horsepower'], axis=1)
display(effi[0:10])

0    2.361538
1    2.121212
2    2.120000
3    2.026667
4    2.157143
5    2.166667
6    2.063636
7    2.046512
8    2.022222
9    2.052632
dtype: float64

### Feature Engineering with Apply and Map

In this section we will see how to calculate a complex feature using map, apply, and grouping.  The data set is the following CSV:

* https://www.irs.gov/pub/irs-soi/16zpallagi.csv 

This is US Government public data for "SOI Tax Stats - Individual Income Tax Statistics".  The primary website is here:

* https://www.irs.gov/statistics/soi-tax-stats-individual-income-tax-statistics-2016-zip-code-data-soi 

Documentation describing this data is at the above link.

For this feature, we will attempt to estimate the adjusted gross income (AGI) for each of the zipcodes.  The data file contains many columns; however, you will only use the following:

* STATE - The state (e.g. MO)
* zipcode - The zipcode (e.g. 63017)
* agi_stub - Six different brackets of annual income (1 through 6) 
* N1 - The number of tax returns for each of the agi_stubs

Note, the file will have 6 rows for each zipcode, for each of the agi_stub brackets. You can skip zipcodes with 0 or 99999.

We will create an output CSV with these columns; however, only one row per zip code. Calculate a weighted average of the income brackets. For example, the following 6 rows are present for 63017:


|zipcode |agi_stub | N1 |
|--|--|-- |
|63017	 |1 | 4710 |
|63017	 |2 | 2780 |
|63017	 |3 | 2130 |
|63017	 |4 | 2010 |
|63017	 |5 | 5240 |
|63017	 |6 | 3510 |


We must combine these six rows into one.  For privacy reasons, AGI's are broken out into 6 buckets.  We need to combine the buckets and estimate the actual AGI of a zipcode. To do this, consider the values for N1:

* 1 = \$1 to \$25,000
* 2 = \$25,000 to \$50,000
* 3 = \$50,000 to \$75,000
* 4 = \$75,000 to \$100,000
* 5 = \$100,000 to \$200,000
* 6 = \$200,000 or more

The median of each of these ranges is approximately:

* 1 = \$12,500
* 2 = \$37,500
* 3 = \$62,500 
* 4 = \$87,500
* 5 = \$112,500
* 6 = \$212,500

Using this you can estimate 63017's average AGI as:

```
>>> totalCount = 4710 + 2780 + 2130 + 2010 + 5240 + 3510
>>> totalAGI = 4710 * 12500 + 2780 * 37500 + 2130 * 62500 + 2010 * 87500 + 5240 * 112500 + 3510 * 212500
>>> print(totalAGI / totalCount)

88689.89205103042
```

In [42]:
import pandas as pd

df=pd.read_csv('https://www.irs.gov/pub/irs-soi/16zpallagi.csv')

First, we trim all zipcodes that are either 0 or 99999.  We also select the three fields that we need.

In [43]:
df=df.loc[(df['zipcode']!=0) & (df['zipcode']!=99999),['STATE','zipcode','agi_stub','N1']]

We replace all of the **agi_stub** values with the correct median values with the **map** function.

In [44]:
medians = {1:12500,2:37500,3:62500,4:87500,5:112500,6:212500}
df['agi_stub']=df.agi_stub.map(medians)

Next the dataframe is grouped by zip code.

In [45]:
groups = df.groupby(by='zipcode')

A lambda is applied across the groups and the AGI estimate is calculated.

In [46]:
df = pd.DataFrame(groups.apply(lambda x:sum(x['N1']*x['agi_stub'])/sum(x['N1']))).reset_index()

The new agi_estimate column is renamed.

In [47]:
df.columns = ['zipcode','agi_estimate']

In [48]:
display(df[0:10])

zipcode  agi_estimate
0     1001  52895.322940
1     1002  64528.451001
2     1003  15441.176471
3     1005  54694.092827
4     1007  63654.353562
5     1008  57575.757576
6     1009  45576.923077
7     1010  61303.191489
8     1011  49807.692308
9     1012  53214.285714

We can also see that our zipcode of 63017 gets the correct value.

In [49]:
df[ df['zipcode']==63017 ]

zipcode  agi_estimate
19909    63017  88689.892051

# Part 2.5: Feature Engineering

Feature engineering is a very important part of machine learning.  Later in this course we will see some techniques for automatic feature engineering.  

## Calculated Fields

It is possible to add new fields to the dataframe that are calculated from the other fields.  We can create a new column that gives the weight in kilograms.  The equation to calculate a metric weight, given a weight in pounds is:

$ m_{(kg)} = m_{(lb)} \times 0.45359237 $

This can be used with the following Python code:

In [50]:
import os
import pandas as pd

df = pd.read_csv(
    "https://data.heatonresearch.com/data/t81-558/auto-mpg.csv", 
    na_values=['NA', '?'])

df.insert(1, 'weight_kg', (df['weight'] * 0.45359237).astype(int))
df

mpg  weight_kg  cylinders  displacement  horsepower  weight  \
0    18.0       1589          8         307.0       130.0    3504   
1    15.0       1675          8         350.0       165.0    3693   
2    18.0       1558          8         318.0       150.0    3436   
3    16.0       1557          8         304.0       150.0    3433   
4    17.0       1564          8         302.0       140.0    3449   
5    15.0       1969          8         429.0       198.0    4341   
6    14.0       1974          8         454.0       220.0    4354   
7    14.0       1955          8         440.0       215.0    4312   
8    14.0       2007          8         455.0       225.0    4425   
9    15.0       1746          8         390.0       190.0    3850   
10   15.0       1616          8         383.0       170.0    3563   
11   14.0       1637          8         340.0       160.0    3609   
12   15.0       1705          8         400.0       150.0    3761   
13   14.0       1399          8         455.0       225.0    3086   
14   24.0       1075          4         113.0        95.0    2372   
15   22.0       1285          6         198.0        95.0    2833   
16   18.0       1258          6         199.0        97.0    2774   
17   21.0       1173          6         200.0        85.0    2587   
18   27.0        966          4          97.0        88.0    2130   
19   26.0        832          4          97.0        46.0    1835   
20   25.0       1211          4         110.0        87.0    2672   
21   24.0       1102          4         107.0        90.0    2430   
22   25.0       1077          4         104.0        95.0    2375   
23   26.0       1013          4         121.0       113.0    2234   
24   21.0       1201          6         199.0        90.0    2648   
25   10.0       2093          8         360.0       215.0    4615   
26   10.0       1984          8         307.0       200.0    4376   
27   11.0       1987          8         318.0       210.0    4382   
28    9.0       2146          8         304.0       193.0    4732   
29   27.0        966          4          97.0        88.0    2130   
..    ...        ...        ...           ...         ...     ...   
368  27.0       1197          4         112.0        88.0    2640   
369  34.0       1086          4         112.0        88.0    2395   
370  31.0       1168          4         112.0        85.0    2575   
371  29.0       1145          4         135.0        84.0    2525   
372  27.0       1240          4         151.0        90.0    2735   
373  24.0       1299          4         140.0        92.0    2865   
374  23.0       1376          4         151.0         NaN    3035   
375  36.0        898          4         105.0        74.0    1980   
376  37.0        918          4          91.0        68.0    2025   
377  31.0        893          4          91.0        68.0    1970   
378  38.0        963          4         105.0        63.0    2125   
379  36.0        963          4          98.0        70.0    2125   
380  36.0        979          4         120.0        88.0    2160   
381  36.0       1000          4         107.0        75.0    2205   
382  34.0       1018          4         108.0        70.0    2245   
383  38.0        891          4          91.0        67.0    1965   
384  32.0        891          4          91.0        67.0    1965   
385  38.0        904          4          91.0        67.0    1995   
386  25.0       1335          6         181.0       110.0    2945   
387  38.0       1367          6         262.0        85.0    3015   
388  26.0       1172          4         156.0        92.0    2585   
389  22.0       1285          6         232.0       112.0    2835   
390  32.0       1208          4         144.0        96.0    2665   
391  36.0       1075          4         135.0        84.0    2370   
392  27.0       1338          4         151.0        90.0    2950   
393  27.0       1265          4         140.0        86.0    2790   
394 

## Google API Keys

Sometimes you will use external API's to obtain data.  The following examples show how to use the Google API keys to encode addresses for use with neural networks.  To use these, you will need your own Google API key.  The key I have below is not a real key, you need to put your own in there.  Google will ask for a credit card, but unless you use a very large number of lookups, there will be no actual cost.  YOU ARE NOT required to get an Google API key for this class, this only shows you how.  If you would like to get a Google API key, visit this site and obtain one for **geocode**.

[Google API Keys](https://developers.google.com/maps/documentation/embed/get-api-key)

In [51]:
GOOGLE_KEY = 'INSERT_YOUR_KEY'

# Other Examples: Dealing with Addresses

Addresses can be difficult to encode into a neural network.  There are many different approaches, and you must consider how you can transform the address into something more meaningful.  Map coordinates can be a good approach.  [Latitude and longitude](https://en.wikipedia.org/wiki/Geographic_coordinate_system) can be a useful encoding.  Thanks to the power of the Internet, it is relatively easy to transform an address into its latitude and longitude values.  The following code determines the coordinates of [Washington University](https://wustl.edu/):

In [52]:
import requests

address = "1 Brookings Dr, St. Louis, MO 63130"

response = requests.get('https://maps.googleapis.com/maps/api/geocode/json?key={}&address={}'.format(GOOGLE_KEY,address))

resp_json_payload = response.json()

if 'error_message' in resp_json_payload:
    print(resp_json_payload['error_message'])
else:
    print(resp_json_payload['results'][0]['geometry']['location'])

{'lat': 38.648238, 'lng': -90.30487459999999}


If latitude and longitude are simply fed into the neural network as two features, they might not be overly helpful.  These two values would allow your neural network to cluster locations on a map.  Sometimes cluster locations on a map can be useful.  Consider the percentage of the population that smokes in the USA by state:

![Smokers by State](https://raw.githubusercontent.com/jeffheaton/t81_558_deep_learning/master/images/class_6_smokers.png "Smokers by State")

The above map shows that certain behaviors, like smoking, can be clustered by global region. 

However, often you will want to transform the coordinates into distances.  It is reasonably easy to estimate the distance between any two points on Earth by using the [great circle distance](https://en.wikipedia.org/wiki/Great-circle_distance) between any two points on a sphere:

The following code implements this formula:

$\Delta\sigma=\arccos\bigl(\sin\phi_1\cdot\sin\phi_2+\cos\phi_1\cdot\cos\phi_2\cdot\cos(\Delta\lambda)\bigr)$

$d = r \, \Delta\sigma$

In [53]:
from math import sin, cos, sqrt, atan2, radians

# Distance function
def distance_lat_lng(lat1,lng1,lat2,lng2):
    # approximate radius of earth in km
    R = 6373.0

    # degrees to radians (lat/lon are in degrees)
    lat1 = radians(lat1)
    lng1 = radians(lng1)
    lat2 = radians(lat2)
    lng2 = radians(lng2)

    dlng = lng2 - lng1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlng / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    return R * c

# Find lat lon for address
def lookup_lat_lng(address):
    response = requests.get('https://maps.googleapis.com/maps/api/geocode/json?key={}&address={}'.format(GOOGLE_KEY,address))
    json = response.json()
    if len(json['results']) == 0:
        print("Can't find: {}".format(address))
        return 0,0
    map = json['results'][0]['geometry']['location']
    return map['lat'],map['lng']


# Distance between two locations

import requests

address1 = "1 Brookings Dr, St. Louis, MO 63130" 
address2 = "3301 College Ave, Fort Lauderdale, FL 33314"

lat1, lng1 = lookup_lat_lng(address1)
lat2, lng2 = lookup_lat_lng(address2)

print("Distance, St. Louis, MO to Ft. Lauderdale, FL: {} km".format(
        distance_lat_lng(lat1,lng1,lat2,lng2)))

Distance, St. Louis, MO to Ft. Lauderdale, FL: 1684.9161446533758 km


Distances can be useful to encode addresses as.  You must consider what distance might be useful for your dataset.  Consider:

* Distance to major metropolitan area
* Distance to competitor
* Distance to distribution center
* Distance to retail outlet

The following code calculates the distance between 10 universities and washu:

In [54]:
# Encoding other universities by their distance to Washington University

schools = [
    ["Princeton University, Princeton, NJ 08544", 'Princeton'],
    ["Massachusetts Hall, Cambridge, MA 02138", 'Harvard'],
    ["5801 S Ellis Ave, Chicago, IL 60637", 'University of Chicago'],
    ["Yale, New Haven, CT 06520", 'Yale'],
    ["116th St & Broadway, New York, NY 10027", 'Columbia University'],
    ["450 Serra Mall, Stanford, CA 94305", 'Stanford'],
    ["77 Massachusetts Ave, Cambridge, MA 02139", 'MIT'],
    ["Duke University, Durham, NC 27708", 'Duke University'],
    ["University of Pennsylvania, Philadelphia, PA 19104", 'University of Pennsylvania'],
    ["Johns Hopkins University, Baltimore, MD 21218", 'Johns Hopkins']
]

lat1, lng1 = lookup_lat_lng("1 Brookings Dr, St. Louis, MO 63130")

for address, name in schools:
    lat2,lng2 = lookup_lat_lng(address)
    dist = distance_lat_lng(lat1,lng1,lat2,lng2)
    print("School '{}', distance to wustl is: {}".format(name,dist))

School 'Princeton', distance to wustl is: 1354.4748428037537
School 'Harvard', distance to wustl is: 1670.6348910867227
School 'University of Chicago', distance to wustl is: 418.07123096093096
School 'Yale', distance to wustl is: 1508.209168740192
School 'Columbia University', distance to wustl is: 1418.2846378506144
School 'Stanford', distance to wustl is: 2780.6884662205066
School 'MIT', distance to wustl is: 1672.4354422735219
School 'Duke University', distance to wustl is: 1046.7924543575177
School 'University of Pennsylvania', distance to wustl is: 1307.1873732319766
School 'Johns Hopkins', distance to wustl is: 1184.3754484499111


# Module 2 Assignment

You can find the first assignment here: [assignment 2](https://github.com/jeffheaton/t81_558_deep_learning/blob/master/assignments/assignment_yourname_class2.ipynb)